Feature/action plugins

targuan opened the pull request:
Since ansible 2.3, network core modules use the same credentials as the others modules to connect to the remote device.
Host, username, password and timeout are no longer required arguments. The provider arguments is set with the default values of these variables.
* host is filled with the current task host
* username is filled with connection_user, which may have been set via an environnement variable, with -u ansible-playbook switch or in ansible.cfg
* password is filled with current password, settable with -k switch

This PR mimics this behaviour for ansible greater or egal to ansible 2.3.0.0

Note that it breaks napalm's compatibility with ansible version lower than 2.2. To use the modules with older version of ansible, the directory action_plugins must be deleted.

dbarrosop commented:
LGTM my only comment is that I'd add to the README a note regarding this:

"Note that it breaks napalm's compatibility with ansible version lower than 2.2. To use the modules with older version of ansible, the directory action_plugins must be deleted."

Something like "if you see this error [error that ansible<2.2 raises] make verify which version of ansible you are running with command blah. If ansible version is 2.2 or lower we recommend upgrading but as an alternative you can try to removing folder blah."

Also, a note on the README on how this works might be interesting (or updated examples on the modules) :D

@ktbyers thoughts?

dbarrosop commented:
Would you mind rebasing? I added some stuff to help with the deployment #79 and versioning so I wanted to take advantage of it. Also, you might need to update https://github.com/napalm-automation/napalm-ansible/blob/develop/napalm_ansible/__init__.py to give instructions on how to use this feature and/or disable it if ansible is very old (you could even check the version yourself and issue a warning).

Also, we added a `CHANGELOG.rst` and a check, if the branch doesn't update it we fail the CI.

Sorry for the mess but I was waiting on this change before merging yours because I thought it could really benefit of it.

targuan commented:
Ok, I think this is good to go.
I rebased the commit to the develop branch, after all the modifications done there. I updated the README, the CHANGELOG and the \_\_init__.py accordingly.
I wasn't shure where to put the action_plugins directory in the hierarchy, but it was settled by the fact that inside napalm_ansible it broke the modules.

ktbyers commented:
Okay, I will try to test and review this.

dbarrosop commented:
@targuan for systems that doesn't support `action_plugins`, it is still necessary to remove the folder or just removing `action_plugins = {action_path}` from the configuration is enough?

dbarrosop commented:
I take from your last commit that the answer is "removing action_plugins = {action_path} from the configuration is enough". Awesome :D

targuan commented:
Actually I wasn't sure it would work for all setup, but after some tests, for the very old version of ansible, napalm_ansible is already not working.
In any case, it will work with the action_plugins loaded for ansible 2.1 to ansible 2.3, even if it only change the behaviour of the modules with ansible 2.3.
Ansible in version 2.0 and lower will fail if the action_plugins are loaded but so does the modules (with the error ``error importing module in /home/targuan/napalm-ansible/napalm_ansible/napalm_get_facts.py, expecting format like 'from ansible.module_utils.<lib name> import *'``.

dbarrosop commented:
@targuan I think we should merge this. Sorry about this but would you mind resubmitting the PR? your branch looks quite broken :S

dbarrosop commented:
I managed to fix your PR here rebasing and squashing but that removed your authorship :(

https://github.com/napalm-automation/napalm-ansible/compare/targuan-feature/action_plugins?expand=1

targuan commented:
I couldn't care less about the authorship. I'll close this PR and let you create another one with your branch.

I really need to work on my git skills ;)
